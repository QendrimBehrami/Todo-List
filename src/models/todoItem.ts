import { v4 as uuidv4 } from "uuid";
import { ITodoItem, Priority } from "../interfaces/todo.interface";

export class TodoItem implements ITodoItem {
  private _id: string;
  private _title: string;
  private _description: string;
  private _dueDate: Date;
  private _priority: Priority;
  private _completed: boolean;

  /**
   * Create a new TodoItem
   * @param title
   * @param description
   * @param dueDate
   * @param priority
   */
  constructor(
    title: string,
    description: string,
    dueDate: Date,
    priority: Priority = Priority.Medium
  ) {
    this._id = uuidv4();
    this._title = title;
    this._description = description;
    this._dueDate = dueDate;
    this._priority = priority;
    this._completed = false;
  }

  get id(): string {
    return this._id;
  }

  get title() {
    return this._title;
  }

  get description() {
    return this._description;
  }

  get dueDate() {
    return this._dueDate;
  }

  get priority() {
    return this._priority;
  }

  set priority(newPriority: Priority) {
    this._priority = newPriority;
  }

  get completed() {
    return this._completed;
  }

  set completed(completed: boolean) {
    this._completed = completed;
  }

  equals(other: TodoItem) {
    return this._id === other._id;
  }
}
