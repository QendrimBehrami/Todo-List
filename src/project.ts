import { v4 as uuidv4 } from "uuid";

export enum Priority {
  Low = "Low",
  Medium = "Medium",
  High = "High",
}

export class TodoItem {
  private _id: string;
  private _title: string;
  private _description: string;
  private _dueDate: Date;
  private _priority: Priority;
  private _completed: boolean;

  constructor(
    title: string,
    description: string,
    dueDate: Date,
    priority: Priority = Priority.Medium,
    id?: string,
    completed = false
  ) {
    this._id = id ?? uuidv4();
    this._title = title;
    this._description = description;
    this._dueDate = dueDate;
    this._priority = priority;
    this._completed = completed;
  }

  get id(): string {
    return this._id;
  }

  get title() {
    return this._title;
  }

  get description() {
    return this._description;
  }

  get dueDate() {
    return this._dueDate;
  }

  get priority() {
    return this._priority;
  }

  set priority(newPriority: Priority) {
    this._priority = newPriority;
  }

  get completed() {
    return this._completed;
  }

  set completed(completed: boolean) {
    this._completed = completed;
  }

  equals(other: TodoItem) {
    return this._id === other._id;
  }
}

export class Project {
  private _todos: TodoItem[];
  private _title: string;
  private _id: string;

  constructor(title: string, todos: TodoItem[] = [], id?: string) {
    this._title = title;
    this._todos = todos ? todos : new Array();
    this._id = id ?? uuidv4();
  }

  public get title(): string {
    return this._title;
  }

  public set title(newTitle) {
    this._title = newTitle;
  }

  public get id() {
    return this._id;
  }

  public addTodo(todo: TodoItem) {
    this._todos.push(todo);
  }

  public removeTodo(todo: TodoItem) {
    this._todos = this._todos.filter(
      (currentTodo) => !currentTodo.equals(todo)
    );
  }

  public get todos(): TodoItem[] {
    return this._todos;
  }
}
