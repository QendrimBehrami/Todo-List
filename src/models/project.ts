import { IProject } from "../interfaces/project.interface";
import { ITodoItem } from "../interfaces/todo.interface";
import { TodoItem } from "./todoItem";

class Project implements IProject {
  private _todos: TodoItem[];
  private _title: string;

  constructor(title: string) {
    this._title = title;
    this._todos = new Array();
  }

  public get title(): string {
    return this._title;
  }

  public set title(newTitle) {
    this._title = newTitle;
  }

  public addTodo(todo: TodoItem) {
    this._todos.push(todo);
  }

  public removeTodo(todo: TodoItem) {
    this._todos = this._todos.filter(
      (currentTodo) => !currentTodo.equals(todo)
    );
  }

  public get todos(): ITodoItem[] {
    return this._todos;
  }
}
