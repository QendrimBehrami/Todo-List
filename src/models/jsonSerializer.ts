import { Project } from "./project";
import { TodoItem } from "./todoItem";
import { ISerializer } from "../interfaces/serializer.interface";

export class JsonProjectSerializer implements ISerializer<Project> {
  serialize(item: Project): string {
    return JSON.stringify(item);
  }

  serializeArray(item: Project[]): string {
    return JSON.stringify(item);
  }

  deserialize(data: string): Project {
    const parsedData = JSON.parse(data);

    const serializer = new JsonTodoItemSerializer();
    let todos = serializer.deserializeArray(JSON.stringify(parsedData._todos));

    return new Project(parsedData._title, todos, parsedData._id);
  }

  deserializeArray(data: string): Project[] {
    const parsedData = JSON.parse(data);

    return parsedData.map((projectData: any) =>
      this.deserialize(JSON.stringify(projectData))
    );
  }
}

export class JsonTodoItemSerializer implements ISerializer<TodoItem> {
  serialize(item: TodoItem): string {
    return JSON.stringify(item);
  }

  serializeArray(item: TodoItem[]): string {
    return JSON.stringify(item);
  }

  deserialize(data: string): TodoItem {
    const parsedData = JSON.parse(data);
    return new TodoItem(
      parsedData._title,
      parsedData._description,
      new Date(parsedData._dueDate),
      parsedData._priority,
      parsedData._id,
      parsedData._completed
    );
  }

  deserializeArray(data: string): TodoItem[] {
    const parsedData = JSON.parse(data);

    return parsedData.map((todoData: any) =>
      this.deserialize(JSON.stringify(todoData))
    );
  }
}
