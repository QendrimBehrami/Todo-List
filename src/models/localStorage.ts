import { ISerializer } from "../interfaces/serializer.interface";
import { IStorage } from "../interfaces/storage.interface";
import { Project } from "./project";

const PROJECT_STORAGE_PREFIX = "PROJECT_";

class LocalProjectStorage implements IStorage<Project> {
  private serializer: ISerializer<Project>;

  constructor(serializer: ISerializer<Project>) {
    this.serializer = serializer;
  }

  save(key: string, item: Project): void {
    const serialized = this.serializer.serialize(item);
    localStorage.setItem(PROJECT_STORAGE_PREFIX + key, serialized);
  }

  load(key: string): Project | null {
    const item = localStorage.getItem(PROJECT_STORAGE_PREFIX + key);
    if (!item) return null;
    return this.serializer.deserialize(item);
  }
}
