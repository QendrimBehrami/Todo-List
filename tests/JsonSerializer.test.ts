import { Priority } from "../src/interfaces/todo.interface";
import { Project } from "../src/models/project";
import { TodoItem } from "../src/models/todoItem";
import { JsonProjectSerializer } from "../src/models/jsonSerializer";

describe("JsonSerializer", () => {
  it("should properly serialize a project to JSON", () => {
    let todoItem = new TodoItem(
      "Write Tests",
      "I need to test the program more...",
      new Date(),
      Priority.High
    );

    let project = new Project("Important Project");
    project.addTodo(todoItem);

    const serializer = new JsonProjectSerializer();
    const serializedString = serializer.serialize(project);
    const deserializedProject = serializer.deserialize(serializedString);

    expect(deserializedProject).toEqual(project);
  });

  it("should properly serialize an array of projects to JSON", () => {
    let todoItem = new TodoItem(
      "Write Tests",
      "I need to test the program more...",
      new Date(),
      Priority.High
    );

    let project = new Project("Important Project");
    project.addTodo(todoItem);

    const serializer = new JsonProjectSerializer();
    const serializedString = serializer.serializeArray([project]);
    console.log(serializedString);

    const deserializedProject = serializer.deserializeArray(serializedString);

    expect(deserializedProject).toEqual([project]);
  });
});
