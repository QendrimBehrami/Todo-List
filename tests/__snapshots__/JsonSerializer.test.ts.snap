// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JsonSerializer should serialize an object to JSON 1`] = `"["{\\"_title\\":\\"Important Project\\",\\"_todos\\":[{\\"_id\\":\\"767bc4c1-6d33-4abf-8118-a4f3b657c1cb\\",\\"_title\\":\\"Write Tests\\",\\"_description\\":\\"I need to test the program more...\\",\\"_dueDate\\":\\"2024-02-25T12:31:38.856Z\\",\\"_priority\\":\\"High\\",\\"_completed\\":false}]}"]"`;
